New MIR optimization pass to reduce branches on match of tuples of enums

simonvandel opened the pull request:
Fixes #68867 by adding a new pass that turns something like
```rust
let x: Option<()>;
let y: Option<()>;
match (x,y) {
    (Some(_), Some(_)) => {0},
    _ => {1}
}
```
into something like
```rust
let x: Option<()>;
let y: Option<()>;
let discriminant_x = // get discriminant of x
let discriminant_y = // get discriminant of x
if discriminant_x != discriminant_y {1} else {0}
```

The opt-diffs still have the old basic blocks like
```
bb3: {
          _8 = discriminant((*(_4.1: &ViewportPercentageLength))); // scope 0 at $DIR/early-otherwise-branch-68867.rs:21:21: 21:30
          switchInt(move _8) -> [1_isize: bb7, otherwise: bb2]; // scope 0 at $DIR/early-otherwise-branch-68867.rs:21:21: 21:30
      }

      bb4: {
          _9 = discriminant((*(_4.1: &ViewportPercentageLength))); // scope 0 at $DIR/early-otherwise-branch-68867.rs:22:23: 22:34
          switchInt(move _9) -> [2_isize: bb8, otherwise: bb2]; // scope 0 at $DIR/early-otherwise-branch-68867.rs:22:23: 22:34
      }

      bb5: {
          _10 = discriminant((*(_4.1: &ViewportPercentageLength))); // scope 0 at $DIR/early-otherwise-branch-68867.rs:23:23: 23:34
          switchInt(move _10) -> [3_isize: bb9, otherwise: bb2]; // scope 0 at $DIR/early-otherwise-branch-68867.rs:23:23: 23:34
      }
```

These do get removed on later passes. I'm not sure if I should include those passes in the test to make it clear?

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @varkor (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

varkor commented:
This looks great, but unfortunately I don't have time to review something of this size at the moment.

r? @eddyb 

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit c6982e36a893aeb5647d1c7773bff287acc43a38 with merge e42444c82f1bdc51f3bd1abadcf3f7b2a9e87159...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c6982e36a893aeb5647d1c7773bff287acc43a38","merge_sha":"e42444c82f1bdc51f3bd1abadcf3f7b2a9e87159"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/942550931), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35763)
Build commit: e42444c82f1bdc51f3bd1abadcf3f7b2a9e87159 (`e42444c82f1bdc51f3bd1abadcf3f7b2a9e87159`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35763","checks-actions":"https://github.com/rust-lang-ci/rust/runs/942550931"},"merge_sha":"e42444c82f1bdc51f3bd1abadcf3f7b2a9e87159"} -->

rust-timer commented:
Queued e42444c82f1bdc51f3bd1abadcf3f7b2a9e87159 with parent 829d69b9c6bfc535a92fc290ec9391a0d5af6c81, future [comparison URL](https://perf.rust-lang.org/compare.html?start=829d69b9c6bfc535a92fc290ec9391a0d5af6c81&end=e42444c82f1bdc51f3bd1abadcf3f7b2a9e87159).

rust-timer commented:
Finished benchmarking try commit (e42444c82f1bdc51f3bd1abadcf3f7b2a9e87159): [comparison url](https://perf.rust-lang.org/compare.html?start=829d69b9c6bfc535a92fc290ec9391a0d5af6c81&end=e42444c82f1bdc51f3bd1abadcf3f7b2a9e87159).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

simonvandel commented:
The pass is gated on Mir opt level 3 right now, so I don't think a perf run makes sense without removing that check. @eddyb do you want me to remove it so we can check the impact of perf?

simonvandel commented:
I don't quite understand how to solve the tests failing.
```
-	                                           // + val: Value(Scalar(0x0000000000000002))
+	                                           // + val: Value(Scalar(0x00000002))
```
Running bless locally does not change anything

Mark-Simulacrum commented:
You likely need to edit the 32-bit tests (which are not run on 64-bit machines) -- I would personally just do so manually. I wish we had a better way to support developers in doing this.

mati865 commented:
You also have to add `// EMIT_MIR_FOR_EACH_BIT_WIDTH` just like in https://github.com/rust-lang/rust/blob/master/src/test/mir-opt/combine_array_len.rs#L1

bors commented:
:umbrella: The latest upstream changes (presumably #73656) made this pull request unmergeable. Please resolve the merge conflicts.

wesleywiser commented:
cc @rust-lang/wg-mir-opt

simonvandel commented:
Rebased on master and squashed

oli-obk commented:
r? @oli-obk

oli-obk commented on src/librustc_mir/transform/early_otherwise_branch.rs:
```diff
@@ -0,0 +1,313 @@
+/// let y: Option<()>;
+/// let discriminant_x = // get discriminant of x
+/// let discriminant_y = // get discriminant of x
+/// if discriminant_x != discriminant_y {1} else {0}
```
isn't it actually `if discriminant_x != discriminant_y || discriminant_x == None`?

lnicola commented on src/librustc_mir/transform/early_otherwise_branch.rs:
```diff
@@ -0,0 +1,313 @@
+/// let x: Option<()>;
+/// let y: Option<()>;
+/// let discriminant_x = // get discriminant of x
+/// let discriminant_y = // get discriminant of x
```
```suggestion
/// let discriminant_y = // get discriminant of y
```

oli-obk commented on src/test/mir-opt/early_otherwise_branch.opt1.EarlyOtherwiseBranch.diff:
```diff
@@ -0,0 +1,88 @@
++         StorageLive(_11);                // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
++         _11 = Ne(_10, _7);               // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
++         StorageDead(_10);                // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
++         switchInt(move _11) -> [false: bb6, otherwise: bb1]; // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
```
instead of doing a switch and going to a block that does another switch, you can also compute another comparison:

```rust
_42 = Eq(_7, 1_isize)
_43 = BitAnd(_42, _11)
```

and then switch on `_43` to directly go to `bb3`

simonvandel replied:
@oli-obk  Hmm that works when we have only have 2 targets in the switch, but what if we have 3 targets in the switch like `match (x,y,z)`? See the 3_element_tuple test. Correct me if I am wrong though

oli-obk replied:
That test makes `bb2` dead code, and doesn't affect the match part on `z` at all, so it all seems to work out to me? You aren't optimizing chains longer than 2 elements with your optimization, and even if the optimization is extended to support that, we'd just end up with more of these `BitAnd`s I think.

simonvandel replied:
It's been a while, but I finally looked at it again. I don't think it is that easy to use BitAnd to avoid the extra bb. See https://github.com/rust-lang/rust/pull/75119/files#diff-37f80fe71480f2b22fb8fd1fc103fda4R88 where we need to check both 0 and 1.

oli-obk commented on src/test/mir-opt/early_otherwise_branch.opt1.EarlyOtherwiseBranch.diff:
```diff
@@ -0,0 +1,88 @@
+- // MIR for `opt1` before EarlyOtherwiseBranch
```
these files (the non-bit-sized ones) should be removed. We don't have a good automatic cleanup procedure for these yet, but are working on something

simonvandel replied:
I removed usize as input as 
a) I was tired of forgetting to update 32bit tests
b) there was no point in testing different bitness

bors commented:
:umbrella: The latest upstream changes (presumably #75370) made this pull request unmergeable. Please resolve the merge conflicts.

crlf0710 commented:
@simonvandel  Ping from triage, could you resolve the merge conflicts? Thanks!

simonvandel commented:
I have rebased on master and added some more commits to cleanup the cfg and add storageDeads for temps that are introduced.
The last commit also enables this pass temporarily on mir-opt-level=1 such that we can check how perf is affected.

Locally, this optimization fires 201 times on rustc-std.

simonvandel commented:
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

oli-obk commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 1d9406eb6eace994260afeb50c5ff9a47af7da8a with merge a1d4724a9a6d3c114437ad892df97967557cd3c6...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1d9406eb6eace994260afeb50c5ff9a47af7da8a","merge_sha":"a1d4724a9a6d3c114437ad892df97967557cd3c6"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1139894011), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36724)
Build commit: a1d4724a9a6d3c114437ad892df97967557cd3c6 (`a1d4724a9a6d3c114437ad892df97967557cd3c6`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36724","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1139894011"},"merge_sha":"a1d4724a9a6d3c114437ad892df97967557cd3c6"} -->

rust-timer commented:
Queued a1d4724a9a6d3c114437ad892df97967557cd3c6 with parent a3bc0e752fad96f537b73f4e9bc805a73d404f7b, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a3bc0e752fad96f537b73f4e9bc805a73d404f7b&end=a1d4724a9a6d3c114437ad892df97967557cd3c6).

rust-timer commented:
Finished benchmarking try commit (a1d4724a9a6d3c114437ad892df97967557cd3c6): [comparison url](https://perf.rust-lang.org/compare.html?start=a3bc0e752fad96f537b73f4e9bc805a73d404f7b&end=a1d4724a9a6d3c114437ad892df97967557cd3c6).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

oli-obk commented:
this is one of the fun perf runs where the overview says we have a regression and the detailed view says we have an improvement :confused:

bjorn3 commented:
Overall this seems to have a positive impact on the max-rss.

oli-obk commented on src/test/mir-opt/early_otherwise_branch_68867.rs:
```diff
@@ -0,0 +1,31 @@
+    Vmax(CSSFloat),
+}
+
+// EMIT_MIR early_otherwise_branch_68867.try_sum.EarlyOtherwiseBranch.diff
```
I think here it would also be very interesting to get a diff from `EarlyOtherwiseBranch` all the way to `SimplifyBranches` after copy prop

simonvandel replied:
I pushed a new commit that adds this (if I understood correctly what you meant)

oli-obk replied:
Ah, no what I meant was a diff between two passes:

```
// EMIT_MIR early_otherwise_branch_68867.try_sum.EarlyOtherwiseBranch.before.mir early_otherwise_branch_68867.try_sum.SimplifyBranches-after-copy-prop.after.mir
```

simonvandel replied:
Rebasing on master, the diff looks even better as `DestinationPropagation` is now in master

oli-obk replied:
But looking at the end result, it doesn't look like we are deduplicating the branches' basic blocks. I guess we need to make branch deduplication smarter for that. We need to teach it that different locals don't mean different logic if the locals are only ever used within that basic block

oli-obk replied:
Can you change the diff so it shows the full multi-optimization diff?

simonvandel replied:
Just pushed a diff from before EarlyOtherwiseBranch to after SimplifyBranches-after-copy-prop. Or do you mean something else?

oli-obk replied:
nope, we just had a race condition :D

simonvandel commented on compiler/rustc_mir/src/transform/early_otherwise_branch.rs:
```diff
@@ -0,0 +1,339 @@
+
+impl<'tcx> MirPass<'tcx> for EarlyOtherwiseBranch {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, source: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        if tcx.sess.opts.debugging_opts.mir_opt_level < 1 {
```
@oli-obk  should I gate this optimization? If yes, which?

oli-obk replied:
Since perf was clean, we can keep it at level 1, so this is fine with me

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 13e8e20df3307d8eaf8b6f86e73f313fcaab39b2 has been approved by `oli-obk`

<!-- @bors r=oli-obk 13e8e20df3307d8eaf8b6f86e73f313fcaab39b2 -->
<!-- homu: {"type":"Approved","sha":"13e8e20df3307d8eaf8b6f86e73f313fcaab39b2","approver":"oli-obk"} -->

simonvandel commented on src/test/mir-opt/early_otherwise_branch_68867.rs:
```diff
@@ -0,0 +1,32 @@
+}
+
+// EMIT_MIR early_otherwise_branch_68867.try_sum.EarlyOtherwiseBranch.diff
+// EMIT_MIR early_otherwise_branch_68867.try_sum EarlyOtherwiseBranch.before SimplifyBranches-after-copy-prop.after
```
@oli-obk Tidy complaines that this is longer than 100 chars. I'm not sure how to compress it. Any ideas?

oli-obk replied:
You add `// ignore-tidy-linelength` to this file

simonvandel replied:
Thanks, pushed that

oli-obk commented:
@bors r-

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 0363694c7ff72d0a4b1c52ebf2320930c3b60da8 has been approved by `oli-obk`

<!-- @bors r=oli-obk 0363694c7ff72d0a4b1c52ebf2320930c3b60da8 -->
<!-- homu: {"type":"Approved","sha":"0363694c7ff72d0a4b1c52ebf2320930c3b60da8","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 0363694c7ff72d0a4b1c52ebf2320930c3b60da8 with merge 2e0edc0f28c5647141bedba02e7a222d3a5dc9c3...
<!-- homu: {"type":"BuildStarted","head_sha":"0363694c7ff72d0a4b1c52ebf2320930c3b60da8","merge_sha":"2e0edc0f28c5647141bedba02e7a222d3a5dc9c3"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1141112114), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36733)
Approved by: oli-obk
Pushing 2e0edc0f28c5647141bedba02e7a222d3a5dc9c3 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36733","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1141112114"},"merge_sha":"2e0edc0f28c5647141bedba02e7a222d3a5dc9c3"} -->

ecstatic-morse commented:
Hi! This PR showed up in the weekly [perf triage](https://github.com/rust-lang/rustc-perf/tree/master/triage) report. It showed mixed results in [instruction counts](https://perf.rust-lang.org/compare.html?start=81e02708f1f4760244756548981277d5199baa9a&end=2e0edc0f28c5647141bedba02e7a222d3a5dc9c3&stat=instructions:u). While `style-servo-opt` seems to have sped up, all other benchmarks run slightly slower. Was this expected?

It's possible that there's some improvement in the generated code, although I'd be a bit surprised if LLVM couldn't manage this. If there's no runtime speedup, I think we should consider disabling this by default.

simonvandel commented:
Hi @ecstatic-morse 
This optimization pass was created in response to #68867 which demonstrated a case where LLVM would not make this optimization.
I did a small comparison (see https://github.com/rust-lang/rust/issues/68867#issuecomment-696975257) which shows that the final asm that LLVM generates looks a lot cleaner. I have not benchmarked what the actualy speedup is in terms of running the function though.

wesleywiser commented:
It sounds like this might be a good fit for mir-opt-level=2 then as (eventually) we want to make that the default for optimized builds while mir-opt-level=1 will remain the default for non optimized builds.

ecstatic-morse commented:
@wesleywiser @simonvandel Was this ever moved to `mir-opt-level=2`? My personal opinion is that new optimizations should go to `mir-opt-level=2` initially unless a) they result in a meaningful performance improvement for most users and b) we are very confident that they don't introduce miscompilations. I don't think this PR fits either of those criteria.

simonvandel commented:
Hi @ecstatic-morse, no this pass is still running at opt-level=1. Should I create a pr to gate it at 2?
